homework/Lesson_16/lesson_16.py

```diff
@@ -0,0 +1,182 @@
+#Task 1
+class Person():
+
+    def __init__(self, firstname, lastname, age):
+        if age < 0:
+            raise ValueError("Age can't be less than 0")
+
+        self.firstname = firstname
+        self.lastname = lastname
+        self.age = age
+
+    def talk(self):
+        print(f"Hello, my name is {self.firstname} {self.lastname} and I'm {self.age} years old")
+
+class Student(Person):
+
+    def __init__(self, firstname, lastname, age, speciality, course):
+        super().__init__(firstname, lastname, age)
+        self.speciality = speciality
+        self.course = course
+
+    def study(self):
+        print(f"""{self.firstname} {self.lastname} is a {self.course} year {self.speciality}. 
+{4 - self.course} years left until  degree""")
+
+class Teacher(Person):
+
+    def __init__(self, firstname, lastname, age, salary, teacher_degree):
+        super().__init__(firstname, lastname, age)
+        self.salary = salary
+        self.teacher_degree = teacher_degree
+
+    def work(self):
+        print(f"{self.firstname} {self.lastname} has a {self.teacher_degree} and receives a salary of  {self.salary} UAH")
+
+
+random_student = Student("Carlie", "Wolf", 20, "student of computer science", 2)
+
+random_teacher = Teacher("Izabela", "Marques", 39, 40000, "doctorate in computer science")
+
+random_teacher.talk()
+random_teacher.work()
+
+random_student.talk()
+random_student.study()
+
+
+
+#Task 2
+class Mathematician(list):
+
+    def square_nums(self, _list):
+        return [i**2 for i in _list]
+
+    def remove_positives(self, _list):
+        return [i for i in _list if i < 0]
+
+    def filter_leaps(self, _list):
+        return [i for i in _list if i % 4 == 0]
+
+
+m = Mathematician()
+
+print(m.square_nums([7, 11, 5, 4]))
+
+print(m.remove_positives([26, -11, -8, 13, -90]))
+
+print(m.filter_leaps([2001, 1884, 1995, 2003, 2020]))
+
+
+
+#Task 3
+class Product:
+    def __init__(self, name, type, price):
+        self.name = name
+        self.type = type
+        self.price = price
+
+
+class Article(Product):
+    def __init__(self, product: Product, amount):
+        self.name = product.name
+        self.amount = amount
+        self.discount = 0
+
+
+class Store:
+    def __init__(self, id):
+        self.id: int = id
+        self.profit: 'percent' = 30
+        self.goods: dict = {}  
+        self.departments: dict = {}
+        self.income = 0
+
+    def add(self, product: Product, amount: int):
+        new = Article(product, amount)
+        new.price = round(product.price + product.price * self.profit/100, 2)
+        self.goods[new.name] = new
+        if product.type not in self.departments.keys():
```

Comment:
It might be a bit complicated to grasp, but as an option, these four lines of code could be rewritten as follows:
```
self.departments[product.type] = self.departments.get(product.type, [])
self.departments[product.type].appent(new)
```
It's debatable if saving two lines of code is worth losing readability, though.